subsys/modbus/modbus_serial.c

```diff
@@ -37,14 +37,31 @@ static void modbus_serial_tx_on(struct modbus_context *ctx)
 		gpio_pin_set(cfg->de->port, cfg->de->pin, 1);
 	}
 
+#if defined CONFIG_MODBUS_SERIAL_IRQ
 	uart_irq_tx_enable(cfg->dev);
+#endif
+
+#if defined CONFIG_MODBUS_SERIAL_ASYNC
+	int ret;
+	uint8_t *buf = cfg->uart_buf;
+	ret = uart_tx(cfg->dev, buf, cfg->uart_buf_ctr, SYS_FOREVER_US);
+
+	if (ret) {
+		LOG_ERR("uart_tx() failed, err %d", ret);
+	}
+	LOG_HEXDUMP_DBG(cfg->uart_buf_ptr, cfg->uart_buf_ctr, "TX data sent:");
+	cfg->uart_buf_ptr += cfg->uart_buf_ctr;
+	cfg->uart_buf_ctr = 0;
```

Comment:
Please move it to modbus_serial_async_start(struct modbus_context *ctx).
modbus_serial_tx_on() should fall back to uart_irq_tx_enable() if CONFIG_UART_ASYNC_API is not enabled.
```c

if (IS_ENABLED(CONFIG_LOG_BACKEND_UART_ASYNC)) {
    modbus_serial_async_start(ctx);
} else {
    uart_irq_tx_enable(cfg->dev);
}
```